Index: include/llvm/Transforms/Instrumentation.h
===================================================================
--- include/llvm/Transforms/Instrumentation.h	(revision 150937)
+++ include/llvm/Transforms/Instrumentation.h	(working copy)
@@ -38,6 +38,9 @@
 // Insert ThreadSanitizer (race detection) instrumentation
 FunctionPass *createThreadSanitizerPass();
 
+// Insert LLCov (live coverage) instrumentation
+ModulePass *createLLCovPass();
+
 } // End llvm namespace
 
 #endif
Index: include/llvm/InitializePasses.h
===================================================================
--- include/llvm/InitializePasses.h	(revision 150937)
+++ include/llvm/InitializePasses.h	(working copy)
@@ -251,6 +251,8 @@
 void initializeUnpackMachineBundlesPass(PassRegistry&);
 void initializeFinalizeMachineBundlesPass(PassRegistry&);
 void initializeBBVectorizePass(PassRegistry&);
+void initializeLLCovPass(PassRegistry&);
+
 }
 
 #endif
Index: lib/Transforms/Instrumentation/Instrumentation.cpp
===================================================================
--- lib/Transforms/Instrumentation/Instrumentation.cpp	(revision 150937)
+++ lib/Transforms/Instrumentation/Instrumentation.cpp	(working copy)
@@ -26,6 +26,7 @@
   initializeGCOVProfilerPass(Registry);
   initializeAddressSanitizerPass(Registry);
   initializeThreadSanitizerPass(Registry);
+  initializeLLCovPass(Registry);
 }
 
 /// LLVMInitializeInstrumentation - C binding for
Index: lib/Transforms/Instrumentation/CMakeLists.txt
===================================================================
--- lib/Transforms/Instrumentation/CMakeLists.txt	(revision 150937)
+++ lib/Transforms/Instrumentation/CMakeLists.txt	(working copy)
@@ -7,4 +7,5 @@
   PathProfiling.cpp
   ProfilingUtils.cpp
   ThreadSanitizer.cpp
+  LLCov.cpp
   )
